pep-0677.rst

```diff
@@ -0,0 +1,851 @@
+our analysis showed they are infrequently used. When they are really
+needed, it is possible to type these using `Callback Protocols
+<https://mypy.readthedocs.io/en/stable/protocols.html#callback-protocols>`_.
+
+See the Rejected Alternatives section for more detailed discussion
+about omitted features.
+
+Specification
+=============
+
+Typing Behavior
+---------------
+
+Type checkers should treat the new syntax with exactly the same
+semantics as ``typing.Callable``.
+
+As such, a type checker should treat the following pairs exactly the
+same::
+
+   from typing import Awaitable, Callable, Concatenate, ParamSpec, TypeVarTuple
+
+    P = ParamSpec("P")
+    Ts = TypeVarTuple('Ts')
+
+    f0: (int, str) -> bool
+    f0: Callable[[int, str], bool]
+
+    f1: (...) -> bool
+    f1: Callable[..., bool]
+
+    f2: async (str) -> str
+    f2: Callable[[str], Awaitable[str]]
+
+    f3: (**P) -> bool
+    f3: Callable[P, bool]
+
+    f4: (int, **P) -> bool
+    f4: Callable[Concatenate[int, P], bool]
+
+    f5: (*Ts) -> bool
+    f5: Callable[[*Ts], bool]
+
+    f6: (int, *Ts, str) -> bool
+    f6: Callable[[int, *Ts, str], bool]
+
+Grammar and Ast
+---------------
+
+The proposed new syntax can be described by these AST changes ::
+
+    expr = <prexisting_expr_kinds>
+         | AsyncCallableType(callable_type_arguments args, expr returns)
+         | CallableType(callable_type_arguments args, expr returns)
+
+    callable_type_arguments = AnyArguments
+                            | ArgumentsList(expr* posonlyargs)
+                            | Concatenation(expr* posonlyargs, expr param_spec)
+
+
+Here are our proposed changes to the `Python Grammar
+<https://docs.python.org/3/reference/grammar.htm>`::
+
+    expression:
+        | disjunction disjunction 'else' expression
+        | callable_type_expression
+        | disjunction
+        | lambdef
+
+    callable_type_expression:
+        | callable_type_arguments '->' expression
+        | ASYNC callable_type_arguments '->' expression
+
+    callable_type_arguments:
+        | '(' '...' [','] ')'
+        | '(' callable_type_positional_argument*  ')'
+        | '(' callable_type_positional_argument* callable_type_param_spec ')'
+
+    callable_type_positional_argument:
+        | !'...' expression ','
+        | !'...' expression &')'
+
+    callable_type_param_spec:
+        | '**' expression ','
+        | '**' expression &')'
+
+
+
+If PEP 646 is accepted, we intend to include support for unpacked
+types by modifying the grammar for
+``callable_type_positional_argument`` as follows::
+
+    callable_type_positional_argument:
+        | expression ','
+        | expression &')'
+        | '*' expression ','
+        | '*' expression &')'
+
+
+Implications of the Grammar
+---------------------------
+
+
+Precedence of ->
+~~~~~~~~~~~~~~~~
+
+
+``->`` binds less tightly than other operators, both inside types and in function signatures::
```

Comment:
Does this mean that the two examples below are equivalent? Would be good to make that explicit.